File: Domain/Project.cs

using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace Domain
{
    public enum TaskStatus
    {
        ToDo,
        InProgress,
        InReview,
        Done
    }

    public enum Visibility
    {
        Public,
        Private
    }

    public enum ProjectStatus
    {
        Active,
        Paused,
    }

    public class Project
    {
        [BsonId]
        [BsonElement("_id")]
        public ObjectId Id { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public int Priority { get; set; }
        public ObjectId OwnerId { get; set; }
        public IList<ObjectId> CollaboratorIds { get; set; } = new List<ObjectId>();
        public DateTime DueDate { get; set; }
        public string Category { get; set; }
        public IList<string> Tags { get; set; } = new List<string>();
        public Visibility Visibility { get; set; }
        public ProjectStatus Status { get; set; }
        public KanbanBoard kanbanBoard { get; set; }
    }

    public class KanbanBoard
    {
        public ObjectId ProjectId { get; set; }
        public string Title { get; set; }
        public IList<ProjectTask> Tasks { get; set; }
    }

    public class ProjectTask
    {
        public ObjectId Id { get; set; }
        public ObjectId ProjectId { get; set; }
        public ObjectId OwnerId { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public DateTime DueDate { get; set; }
        public TaskStatus Status { get; set; }
        public IList<ObjectId> PeopleAssigned { get; set; } = new List<ObjectId>();
    }
}

File: ui/src/app/models/project.ts
/* eslint-disable no-magic-numbers */
import ObjectID from "bson-objectid";
//Models & Types
import { Profile } from "./profile";

export enum Visibility
{
    Public = 0,
    Private = 1,
}

export enum ProjectStatus
{
    Active,
    Paused,
}

export enum TaskStatus
{
    ToDo = "ToDo",
    InProgress = "InProgress",
    InReview = "InReview",
    Done = "Done",
}

export interface Project
{
    id?: ObjectID;
    title: string;
    description: string;
    priority: number;
    owner: Profile;
    collaborators: Profile[];
    dueDate: Date;
    category: string;
    tags: string[];
    visibility: Visibility;
    status: ProjectStatus;
    kanbanBoard: KanbanBoard;
}

export interface KanbanBoard
{
    id?: string;
    projectId?: ObjectID;
    title: string;
    tasks: Task[];
}

export interface Task
{
    id?: ObjectID;
    projectId: ObjectID;
    ownerId: ObjectID;
    name: string;
    description: string;
    dueDate: Date;
    peopleAssigned: ObjectID[];
    status: TaskStatus;
}